docs/DesignDocs/InitTimeConfigurationInterface.md

```diff
@@ -0,0 +1,715 @@
+# Open Enclave Init-time Configuration Interface
+
+In this paper, we present the interface design for Init-time Configuration
+support in Open Enclave. Open Enclave Init-time Configuration feature supports
+configurable enclave functionality beyond the static functionality implemented
+by the enclave code and initialized data inside the enclave at the enclave
+initialization time.
+
+## SGX CONFIGID and CONFIGSVN Overview
+
+Intel's 10th-gen Core processor and 3rd-gen Xeon-SP processor support the SGX
+Key Separation and Sharing (KSS) feature, including `CONFIGID` and `CONFIGSVN`,
+which are new fields defined in `SECS`, set by the host side SW at Enclave
+initialization time. `CONFIGID` and `CONFIGSVN` enable configurable enclave
+functionality, in contrast to the static functionality decided by the enclave
+code and initialized data, which are reflected in SGX `MRENCLAVE` and can not
+change without re-signing the enclave. For example, an enclave app might
+configure the enclave with a public key of an external service the enclave will
+trust, instead of relying on a hardcoded value. Another form of configurable
+enclave functionality is loading additional code/data into the enclave post
+enclave initialization. Enclave functionality change based on the configuration
+data might have security implication, therefore, the configuration data must be
+captured in the enclave attestation. `CONFIGID` and `CONFIGSVN` are part of the
+enclave identity produced by the CPU, reflecting the configuration committed at
+the enclave initialization time.
```

Comment:
An alternate approach is to use layered attestation, where the enclave signs the configuration data, and the evidence chains the two together.  Enclave signs the data, lower layer (CPU for SGX) signs the enclave code, and the evidence includes both.  That's how DICE works for example, and doesn't require any changes to the lower layer (e.g., works with older CPUs).  I have no objection to supporting it in addition as stated in the document, but I'd just like to point out this is very SGX specific and a more general mechanism would work for SGX too, even without any KSS feature, if I understand correctly.

From my previous meetings with Christoph et al., the only time the layered approach is not sufficient is if the next layer is code (not just data) running in the _same_ address space (the Javascript interpreter being the canonical example).  That's what _needs_ an approach whereby the lower layer signs both the enclave and the code to be loaded in the enclave. The paragraph at 37-57 does make the point about the code, and it would be helpful to augment it to say that when the data is only data, not code, that the proposed approach can still work but so would alternate approaches such as layered attestation, which would be needed in other cases as well (like code loaded dynamically into a separate address space, which SGX does not currently support).

Reply:
Agreed. Add a clarification to highlight that immutable identity of loaded code is required because SGX technology does not provide privilege isolation among SW modules within the enclave.